feat: refactor and add protocol support to `stats/base/nanmeanwd`

prajjwalbajpai opened the pull request:
Resolves #5658 .

## Description

> What is the purpose of this pull request?

This pull request:

-  Adds support for accessor arrays and refactor `stats/base/nanmeanwd`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #5658 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

prajjwalbajpai committed:
feat: refactored and added protocol support to stats/base/nanmeanwd

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai commented:
@aayush0325 please review

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/docs/repl.txt:
```diff
@@ -1,10 +1,10 @@
 
-    The `N` and `stride` parameters determine which elements in `x` are accessed
-    at runtime.
+    The `N` and `stride` parameters determine which elements
```
can you please change the line wrapping as we can include more words here before hitting the 80 char limit

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/docs/types/test.ts:
```diff
@@ -26,6 +27,8 @@ import nanmeanwd = require( './index' );
 
 	nanmeanwd( x.length, x, 1 ); // $ExpectType number
+	nanmeanwd( x.length, new AccessorArray( x ), 1 ); // $ExpectType number
+
```
```suggestion
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/docs/types/test.ts:
```diff
@@ -85,6 +88,8 @@ import nanmeanwd = require( './index' );
 
 	nanmeanwd.ndarray( x.length, x, 1, 0 ); // $ExpectType number
+	nanmeanwd.ndarray( x.length, new AccessorArray( x ), 1, 0 ); // $ExpectType number
+
```
```suggestion
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/lib/ndarray.js:
```diff
@@ -43,33 +49,35 @@
 * @example
-* var floor = require( '@stdlib/math/base/special/floor' );
+* var x = [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0, NaN, NaN ];
 *
```
```suggestion
```

aayush0325 commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/test/test.main.js:
```diff
@@ -40,7 +40,7 @@ tape( 'the function has an arity of 3', function test( t ) {
 });
 
-tape( 'the function calculates the arithmetic mean of a strided array', function test( t ) {
+tape( 'the function calculates the arithmetic mean of a strided array (ignoring `NaN` values)', function test( t ) {
```
this change is not needed here

aayush0325 replied:
same comment for all other instances

aayush0325 reviewed (changes requested):
looking good, minor comments

prajjwalbajpai committed:
fix: modified repl.txt and more

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

prajjwalbajpai commented:
> looking good, minor comments

Done

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into refactor-nanmeanwd

gururaj1512 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into refactor-nanmeanwd

gururaj1512 committed:
docs: update README

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
chore: update documentation for nanmeanwd

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: update examples for nanmeanwd

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
test: add accessor tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: update notes in README

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/docs/types/index.d.ts:
```diff
@@ -57,15 +62,15 @@ interface Routine {
 	* // returns ~0.3333
 	*/
-	ndarray( N: number, x: NumericArray, stride: number, offset: number ): number;
+	ndarray( N: number, x: InputArray, strideX: number, offset: number ): number;
```
```suggestion
	ndarray( N: number, x: InputArray, strideX: number, offsetX: number ): number;
```

kgryte committed:
chore: fix parameter name

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/lib/main.js:
```diff
@@ -20,14 +20,30 @@
+* @param {PositiveInteger} N - number of indexed elements
+* @param {NumericArray} x - input array
+* @param {integer} strideX - stride length
+* @returns {number} variance
```
```suggestion
* @returns {number} arithmetic mean
```

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/lib/main.js:
```diff
@@ -20,14 +20,30 @@
+*
+* @example
+* var x = [ 1.0, -2.0, NaN, 2.0 ];
+* var N = x.length;
+*
+* var v = nanmeanwd( x.length, x, 1 );
```
```suggestion
*
* var v = nanmeanwd( x.length, x, 1 );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/lib/ndarray.js:
```diff
@@ -43,33 +49,35 @@
+		return accessors( N, o, strideX, offsetX );
+	}
+	if ( N === 1 || strideX === 0 ) {
+		v = x[ offsetX ];
```
```suggestion
		return x[ offsetX ];
```

kgryte committed:
fix: ensure early return

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/lib/accessors.js:
```diff
@@ -41,45 +41,53 @@
+	get = x.accessors[ 0 ];
+
+	if ( N === 1 || strideX === 0 ) {
+		v = get( xbuf, offsetX );
```
```suggestion
		return get( xbuf, offsetX );
```

kgryte committed:
fix: ensure early return

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/nanmeanwd/lib/accessors.js:
```diff
@@ -41,45 +41,53 @@
+	if ( N === 1 || strideX === 0 ) {
+		return get( xbuf, offsetX );
 	}
+
```
```suggestion
```

kgryte committed:
style: remove empty line

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after clean-up.
